gobblin-data-management/src/main/java/gobblin/data/management/copy/hive/HiveTargetPathHelper.java

```diff
@@ -0,0 +1,185 @@
+        : Optional.<Path> absent();
+  }
+
+  private static Path addPartitionToPath(Path path, Partition partition) {
+    for (String partitionValue : partition.getValues()) {
+      path = new Path(path, partitionValue);
+    }
+    return path;
+  }
+
+  /**
+   * Takes a path with tokens {@link #databaseToken} or {@link #tableToken} and replaces these tokens with the actual
+   * database names and table name. For example, if db is myDatabase, table is myTable, then /data/$DB/$TABLE will be
+   * resolved to /data/myDatabase/myTable.
+   */
+  protected static Path resolvePath(String pattern, String database, String table) {
+    pattern = pattern.replace(databaseToken, database);
+    if (pattern.contains(tableToken)) {
+      pattern = pattern.replace(tableToken, table);
+      return new Path(pattern);
+    } else {
+      return new Path(pattern, table);
+    }
+  }
+
+  /**
+   * Compute the target {@link Path} for a file or directory copied by Hive distcp.
+   *
+   * <p>
+   *   The target locations of data files for this table depend on the values of the resolved table root (e.g.
+   *   the value of {@link #COPY_TARGET_TABLE_ROOT} with tokens replaced) and {@link #RELOCATE_DATA_FILES_KEY}:
+   *   * if {@link #RELOCATE_DATA_FILES_KEY} is true, then origin file /path/to/file/myFile will be written to
+   *     /resolved/table/root/<partition>/myFile
+   *   * if {@link #COPY_TARGET_TABLE_PREFIX_TOBE_REPLACED} and {@link #COPY_TARGET_TABLE_PREFIX_REPLACEMENT} are defined,
+   *     then the specified prefix in each file will be replaced by the specified replacement.
+   *   * otherwise, if the resolved table root is defined (e.g. {@link #COPY_TARGET_TABLE_ROOT} is defined in the
+   *     properties), we define:
+   *     origin_table_root := the deepest non glob ancestor of table.getSc().getLocation() iff getLocation() points to
+   *                           a single glob. (e.g. /path/to/*&#47;files -> /path/to). If getLocation() contains none
+   *                           or multiple globs, job will fail.
+   *     relative_path := path of the file relative to origin_table_root. If the path of the file is not a descendant
+   *                      of origin_table_root, job will fail.
+   *     target_path := /resolved/table/root/relative/path
+   *     This mode is useful when moving a table with a complicated directory structure to a different base directory.
+   *   * otherwise the target is identical to the origin path.
+   * </p>
+   *
+   *
+   * @param sourcePath Source path to be transformed.
+   * @param targetFs target {@link FileSystem}
+   * @param partition partition this file belongs to.
+   * @param isConcreteFile true if this is a path to an existing file in HDFS.
+   */
+  public Path getTargetPath(Path sourcePath, FileSystem targetFs, Optional<Partition> partition, boolean isConcreteFile) {
+    if (this.relocateDataFiles) {
```

Comment:
They do different things. If `COPY_TARGET_TABLE_ROOT` and `RELOCATE_DATA_FILES`, then files will be places directly under the new table root (it will look very much like a Hive managed table), while if `RELOCATE_DATA_FILES` is false, we will respect the relative paths of each file, just replace the old table root by the new table root. If both are false, then we will preserve the paths exactly.